spec/unit/service_spec.rb

```diff
@@ -461,6 +463,60 @@ module Backend
             expect(Service.save!(svc_attrs).active?).to be_falsy
           end
         end
+
+        it 'returns true when the service does not have state in the DB' do
+          service_id = '9001'
+          Service.save!({state: :suspended, id: service_id, default_service: false})
+          Service.storage.del ThreeScale::Backend::Service.storage_key(service_id, 'state')
+          expect(Service.load_by_id(service_id).active?).to be true
+          Service.delete_by_id(service_id)
+        end
+
+        it 'returns false when the service has an invalid state in the DB' do
+          service_id = '9001'
+          Service.save!({state: :not_defined_state, id: service_id, default_service: false})
+          Service.storage.set ThreeScale::Backend::Service.storage_key(service_id, 'state'), 'not_defined_state'
+          expect(Service.load_by_id(service_id).active?).to be false
+          Service.delete_by_id(service_id)
+        end
+      end
+
+      describe '#state' do
```

Comment:
those tests are repetition of ```#active?``` tests

Reply:
They are not. That's because active can return true/false and that's what is checked there, whereas the state getter tests verify what values should be returned.

Reply:
service class public interface should be ```#active?```. 

IMO test plan should only test ```#active?```. They should not test ```state``` accessor, it should even be private accessor, nobody should access it externally

Reply:
I thought you wanted it to be public too as written in the original commit because the attribute_reader is set as public and the setter too.

If your idea was to only have the #active? method as the public interface, which I agree, then we could do the following thing if the style is correct (let me know):

```
diff --git a/lib/3scale/backend/service.rb b/lib/3scale/backend/service.rb
index 9c46aad2..a766f2a9 100644
--- a/lib/3scale/backend/service.rb
+++ b/lib/3scale/backend/service.rb
@@ -10,6 +10,8 @@ module ThreeScale
       private_constant :ATTRIBUTES
 
       attr_reader :state
+      private :state
+
       attr_accessor :provider_key, :id, :backend_version,
         :default_user_plan_id, :default_user_plan_name
       attr_writer :referrer_filters_required, :user_registration_required,
@@ -239,6 +241,10 @@ module ThreeScale
         state == :active
       end
 
+
+      private
+
+
       def state=(value)
         # only :active or nil will be considered as :active
         # we assume nil is active because not having a state in an
@@ -247,8 +253,6 @@ module ThreeScale
         @state = value.nil? || value.to_sym == :active ? :active : :suspended
       end
 
-      private
-
       def delete_attributes
         keys = ATTRIBUTES.map { |attr| storage_key(attr) }
         keys << storage_key(:user_set)
```

Reply:
ok for the attr reader.

The attr writer I would leave it open.

If the future features require *state* to be multivalue (not just binary value), we can open attr reader

Reply:
If we leave the attr writer as open, shouldn't we be able to test in some way how the method should behave?

Reply:
testing ```#active?```

Maybe we should leave ```state``` as private as you suggested and create new public method
```
def set_active(active)
   state = active ? :active : :suspended
end
```

Reply:
So, the idea would be to make tests of the set_active method because it would be public, and then in the test test that the call to active? returns true???

Reply:
the whole idea is an abstraction over service state. The public api allows you to check whether a service is active or not using ```#active?``` method. The public api allows you to update service activeness doing 
```
service.set_disabled
service.set_enabled
```
or 
```
service.set_enabled(false)
service.set_enabled(true)
```

Internal *state* attr and database serialization is implementation detail.

Reply:
I propose the following:

Implement the methods:

```
service.enable
service.disable
```

Those methods would call the private setter, and the getter would be private too.
Tests for those new methods would be implemented, and inside those methods we would check the results of the #active? method after calling them.

What do you think?

Reply:
I have implemented the enable and disable methods

Reply:
Finally I removed the implementation of enable/disable methods because could cause confusion implying that calling them would write to the DB. I created the active=(value) setter instead, having value as a boolean.
Also active has been set as an alias of active?